File: ios/HoloNet/Module/Forum/Data/ForumRepositoryBase.swift
//
//  ForumRepositoryBase.swift
//  SWTOR HoloNet
//
//  Created by Ivan Fabijanovic on 14/07/15.
//  Copyright (c) 2015 Ivan Fabijanović. All rights reserved.
//

import UIKit
import Alamofire
import HTMLReader

enum ForumError: Error {
    case maintenance
}

class ForumRepositoryBase {
    let parser: ForumParser
    let settings: Settings
    
    let manager: Alamofire.SessionManager
    
    init(parser: ForumParser, settings: Settings) {
        self.parser = parser
        self.settings = settings
        
        self.manager = Alamofire.SessionManager.default
    }
    
    func localizedSettings(language: ForumLanguage) -> LocalizedSettings {
        let localizedSettings = self.settings.localized[language.rawValue]
        assert(localizedSettings != nil)
        return localizedSettings!
    }
}

extension ForumRepositoryBase {
    func isMaintenanceResponse(_ html: String) -> Bool {
        let document = HTMLDocument(string: html)
        let errorNodes = document.nodes(matchingSelector: "#mainContent > #errorPage #errorBody p")
        
        if let englishNode = errorNodes.first {
            return englishNode.textContent.range(of: "scheduled maintenance") != nil
        }
        
        return false
    }
}

File: android/app/src/main/java/com/ifabijanovic/holonet/forum/data/ForumRepositoryBase.kt
package com.ifabijanovic.holonet.forum.data

import com.ifabijanovic.holonet.app.model.LocalizedSettings
import com.ifabijanovic.holonet.app.model.Settings
import com.ifabijanovic.holonet.forum.language.ForumLanguage
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory

/**
 * Created by feb on 19/03/2017.
 */
abstract class ForumRepositoryBase(
        protected val parser: ForumParser,
        protected val service: ForumService,
        protected val settings: Settings
) {
    protected fun localizedSettings(language: ForumLanguage): LocalizedSettings {
        val localizedSettings = settings.localized[language.languageCode]
        assert(localizedSettings != null)
        return localizedSettings!!
    }

    protected fun isMaintenance(html: String): Boolean {
        val document = Jsoup.parse(html)
        val errorNodes = document.select("#mainContent > #errorPage #errorBody p")

        if (errorNodes == null || errorNodes.isEmpty()) { return false }
        // TODO implement based on language
        return errorNodes.first().text().contains("scheduled maintenance", true)
    }
}
